app/plugin.go

```diff
@@ -182,6 +190,78 @@ func (a *App) InitPlugins(pluginDir, webappPluginDir string) {
 	a.SyncPluginsActiveState()
 }
 
+// SynchPlugins will synchronize the plugins installed locally
+// with the plugin bundles available in the file store.
+func (a *App) SynchPlugins() *model.AppError {
+	mlog.Info("Synching plugins with the file store ...")
+
+	pluginDirs, err := ioutil.ReadDir(*a.Config().PluginSettings.Directory)
+	if err != nil {
+		return model.NewAppError("SynchPlugins", "app.plugin.synch.read_local_folder.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// Remove plugins locally to synch with file store.
+	for _, dir := range pluginDirs {
+		if !dir.IsDir() {
+			continue
+		}
+		dirFullPath := filepath.Join(*a.Config().PluginSettings.Directory, dir.Name())
+		files, readErr := ioutil.ReadDir(dirFullPath)
+
+		if readErr != nil {
+			mlog.Error(fmt.Sprintf("Error reading local plugin directoy: %v. Skipped for synch.", dir.Name()), mlog.Err(readErr))
+			continue
+		}
+		// Only handle managed plugins with .filestore flag file.
+		managed := false
+		for _, f := range files {
+			if f.Name() == ".filestore" {
+				managed = true
+
+				mlog.Debug(fmt.Sprintf("Plugin Synch: Uninstalling %v plugin locally", dir.Name()))
+				if err := a.removePlugin(dir.Name(), true); err != nil {
+					mlog.Error(fmt.Sprintf("Plugin Synch: Error uninstalling managed plugin: %v.", dir.Name()), mlog.Err(err))
+				}
+				break
+			}
+		}
+
+		if !managed {
+			mlog.Warn(fmt.Sprintf("Found unmanaged plugin: %v. Ignoring in plugin synch with the filestore.", dir.Name()))
+		}
+	}
+
+	// Install plugins from the file store.
+	exists, existsErr := a.FileExists("./plugins")
+
+	if existsErr != nil {
+		return model.NewAppError("SynchPlugins", "app.plugin.synch.check_filestore.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	if exists {
+		fileStorepaths, listDirErr := a.ListAll("./plugins")
+		if listDirErr != nil {
+			return model.NewAppError("SynchPlugins", "app.plugin.synch.list_filestore.app_error", nil, listDirErr.Error(), http.StatusInternalServerError)
+		}
+
+		for _, path := range fileStorepaths {
+			if strings.HasSuffix(path, ".tar.gz") {
+				fileBytes, fileReaderErr := a.ReadFile(filepath.Join("./", path))
+				if fileReaderErr != nil {
+					mlog.Error(fmt.Sprintf("Failed to open plugin bundle from filestore: %v", path), mlog.Err(fileReaderErr))
+					continue
+				}
+
+				mlog.Debug(fmt.Sprintf("Plugin Synch: installing %v plugin locally", path))
+				if _, err := a.InstallPlugin(bytes.NewReader(fileBytes), true); err != nil {
```

Comment:
Bit worried about calling install plugin directly from here. It will probably do what you need for now. But I think we should separate the "admin has installed a plugin" path from the "a plugin has been installed from the file store" path. Otherwise we might end up with some strange behavior.

Reply:
Refactored into:

`installPlugin`: High level method that does all operations for installing the plugin locally and any synch operations with the filestore (and will include cluster notifications when merged with the other [PR](https://github.com/mattermost/mattermost-server/pull/11611))
`installPluginLocally`: Performs all local operations to install the plugin